File: src/TimeFormat.cpp
#include "TimeFormat.h"

#include <QRegularExpression>
#include <QtMath>

namespace TimeFormat {

QString formatMs(qint64 milliseconds, bool includeMillis)
{
    if (milliseconds < 0)
        milliseconds = 0;

    const qint64 totalMs = milliseconds;
    const qint64 hours = totalMs / 3600000;
    const qint64 minutes = (totalMs % 3600000) / 60000;
    const qint64 seconds = (totalMs % 60000) / 1000;
    const qint64 millis = totalMs % 1000;

    if (includeMillis) {
        return QStringLiteral("%1:%2:%3.%4")
            .arg(hours, 2, 10, QLatin1Char('0'))
            .arg(minutes, 2, 10, QLatin1Char('0'))
            .arg(seconds, 2, 10, QLatin1Char('0'))
            .arg(millis, 3, 10, QLatin1Char('0'));
    }

    return QStringLiteral("%1:%2:%3")
        .arg(hours, 2, 10, QLatin1Char('0'))
        .arg(minutes, 2, 10, QLatin1Char('0'))
        .arg(seconds, 2, 10, QLatin1Char('0'));
}

qint64 parseToMs(const QString& text, bool* ok)
{
    const QString trimmed = text.trimmed();
    static const QRegularExpression re(
        QStringLiteral(R"(^(?:(\d+):)?(\d{1,2}):(\d{1,2})(?:\.(\d{1,3}))?$)"));

    const auto match = re.match(trimmed);
    if (!match.hasMatch()) {
        if (ok)
            *ok = false;
        return 0;
    }

    const qint64 hours = match.captured(1).isEmpty() ? 0 : match.captured(1).toLongLong();
    const qint64 minutes = match.captured(2).toLongLong();
    const qint64 seconds = match.captured(3).toLongLong();
    QString millisStr = match.captured(4);
    if (millisStr.isEmpty())
        millisStr = QStringLiteral("0");
    while (millisStr.size() < 3)
        millisStr.append(QLatin1Char('0'));

    const qint64 millis = millisStr.left(3).toLongLong();
    if (ok)
        *ok = true;
    return hours * 3600000 + minutes * 60000 + seconds * 1000 + millis;
}

} // namespace TimeFormat

File: Trimmi.Core/Services/TimeFormat.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Trimmi.Core.Services;

public static partial class TimeFormat
{
    [GeneratedRegex(@"^(?:(\d+):)?(\d{1,2}):(\d{1,2})(?:\.(\d{1,3}))?$", RegexOptions.CultureInvariant)]
    private static partial Regex TimeRegex();

    public static string FormatMs(long ms, bool includeMillis = true)
    {
        if (ms < 0)
            ms = 0;

        var hours = ms / 3_600_000;
        var minutes = (ms % 3_600_000) / 60_000;
        var seconds = (ms % 60_000) / 1000;
        var millis = ms % 1000;

        if (includeMillis)
            return $"{hours:D2}:{minutes:D2}:{seconds:D2}.{millis:D3}";

        return $"{hours:D2}:{minutes:D2}:{seconds:D2}";
    }

    public static bool TryParseToMs(string text, out long ms)
    {
        ms = 0;
        var trimmed = text.Trim();
        var match = TimeRegex().Match(trimmed);
        if (!match.Success)
            return false;

        var hours = match.Groups[1].Success
            ? long.Parse(match.Groups[1].Value, CultureInfo.InvariantCulture)
            : 0;
        var minutes = long.Parse(match.Groups[2].Value, CultureInfo.InvariantCulture);
        var seconds = long.Parse(match.Groups[3].Value, CultureInfo.InvariantCulture);

        var millisStr = match.Groups[4].Success ? match.Groups[4].Value : "0";
        while (millisStr.Length < 3)
            millisStr += "0";
        if (millisStr.Length > 3)
            millisStr = millisStr[..3];

        var millis = long.Parse(millisStr, CultureInfo.InvariantCulture);
        ms = hours * 3_600_000 + minutes * 60_000 + seconds * 1000 + millis;
        return true;
    }
}
